sklearn/datasets/_arff_parser.py

```diff
@@ -69,176 +69,212 @@ def _sparse_data_to_array(
-                if cat is not None and not is_scalar_nan(cat)
-            ]
-            dtype = pd.api.types.CategoricalDtype(cats_without_missing)
-        df[column] = df[column].astype(dtype, copy=False)
-    return (df,)
+    y : {series, dataframe} or None
+        The series or dataframe containing the target.
+    """
+    X = frame[feature_names]
+    if len(target_names) >= 2:
+        y = frame[target_names]
+    elif len(target_names) == 1:
+        y = frame[target_names[0]]
+    else:
+        y = None
+    return X, y
 
 
 def _liac_arff_parser(
-    arff_container,
+    gzip_file,
     output_arrays_type,
-    features_dict,
-    data_columns,
-    target_columns,
-    col_slice_x=None,
-    col_slice_y=None,
+    columns_info_openml,
+    feature_names_to_select,
+    target_names_to_select,
     shape=None,
 ):
+    """ARFF parser using the LIAC-ARFF library coded purely in Python.
+
+    This parser is quite slow but consumes a generator.
+
+    Parameters
+    ----------
+    gzip_file : GzipFile instance
+        The file compressed to be read.
+
+    output_array_type : {"numpy", "sparse", "pandas"}
+        The type of the arrays that will be returned. The possibilities ara:
+
+        - `"numpy"`: both `X` and `y` will be NumPy arrays;
+        - `"sparse"`: `X` will be sparse matrix and `y` will be a NumPy array;
+        - `"pandas"`: `X` will be a pandas DataFrame and `y` will be either a
+          pandas Series or DataFrame.
+
+    columns_info : dict
+        The information provided by OpenML regarding the columns of the ARFF
+        file.
+
+    feature_names_to_select : list of str
+        A list of the feature names to be selected.
+
+    target_names_to_select : list of str
+        A list of the target names to be selected.
+
+    Returns
+    -------
+    X : {ndarray, sparse matrix, dataframe}
+        The data matrix.
+
+    y : {ndarray, dataframe, series}
+        The target.
+
+    frame : dataframe or None
+        A dataframe containing both `X` and `y`. `None` if
+        `output_array_type != "pandas"`.
+
+    nominal_attributes : list of str or None
+        The names of the features that are categorical. `None` if
+        `output_array_type == "pandas"`.
+    """
+
+    def _io_to_generator(gzip_file):
+        for line in gzip_file:
+            yield line.decode("utf-8")
+
+    stream = _io_to_generator(gzip_file)
+
+    # find which type (dense or sparse) ARFF type we will have to deal with
+    return_type = _arff.COO if output_arrays_type == "sparse" else _arff.DENSE_GEN
+    # we should not let LIAC-ARFF to encode the nominal attributes with NumPy
+    # arrays to have only numerical values.
+    encode_nominal = not (output_arrays_type == "pandas")
+    arff_container = _arff.load(
+        stream, return_type=return_type, encode_nominal=encode_nominal
+    )
+    columns_to_select = feature_names_to_select + target_names_to_select
+
+    nominal_attributes = {
```

Comment:
Maybe this should be renamed to "categorical" because we do not really know if categorical variables should be treated as nominal or ordinal in ARFF (e.g. for categories encoded with integer values).